File: nodejs/RtcTokenBuilder.js
'use strict'

const zlib = require('zlib');
const crypto = require('crypto');

const Base64URL = require('./Base64URL');

class RtcTokenBuilder {

  constructor(appId, secret) {
    this.appId = appId;
    this.secret = secret;
  }

  buildToken(roomId, userId, expire) {
    const time = Date.now() / 1000 | 0;
    const nonce = getNonce(5);
    const signDoc = {
      ver: '1.0',
      nonce,
      time,
      expire
    }

    const signature = crypto.createHmac('sha256', this.secret).update(`${this.appId}:${roomId}:${userId}:${nonce}:${time}:${expire}`).digest('base64');
    signDoc.signature = signature;
    return Base64URL.escape(zlib.deflateSync(Buffer.from(JSON.stringify(signDoc))).toString('base64'));
  }

}

module.exports = RtcTokenBuilder;

function getNonce(chars) {
  const d = []
  for (let i = 0; i < chars; i++) {
    d.push(Math.round(Math.random() * 10))
  }
  return d.join('')
}

File: python/RtcTokenBuilder.py
#! /usr/bin/python
# coding:utf-8


import hmac
import hashlib
import base64
import zlib
import json
import time
import random


def base64_encode_url(data):
  """ base url encode 实现"""
  base64_data = base64.b64encode(data)
  base64_data_str = bytes.decode(base64_data)
  base64_data_str = base64_data_str.replace('+', '*')
  base64_data_str = base64_data_str.replace('/', '-')
  base64_data_str = base64_data_str.replace('=', '_')
  return base64_data_str


def base64_decode_url(base64_data):
  """ base url decode 实现"""
  base64_data_str = bytes.decode(base64_data)
  base64_data_str = base64_data_str.replace('*', '+')
  base64_data_str = base64_data_str.replace('-', '/')
  base64_data_str = base64_data_str.replace('_', '=')
  raw_data = base64.b64decode(base64_data_str)
  return raw_data


class RtcTokenBuilder:
  __appId = ""
  __secret = ""

  def __init__(self, appId, secret):
    self.__appId = appId
    self.__secret = secret

  def __hmacsha256(self, roomId, userId, nonce, currTime, expire):
    """ 通过固定串进行 hmac 然后 base64 得的 sig 字段的值"""
    raw_content_to_be_signed = self.__appId + ":" + roomId + ":" + userId + ":" + nonce + ":" + str(currTime) + ":" + str(expire)
    return base64.b64encode(hmac.new(self.__secret.encode('utf-8'),
                                      raw_content_to_be_signed.encode('utf-8'),
                                      hashlib.sha256).digest())

  def buildToken(self, roomId, userId, expire):
    """ 用户可以采用默认的有效期生成 sig """
    nonce = str(random.random() * 1000000)
    curr_time = int(time.time())
    m = dict()
    m["ver"] = "1.0"
    m["nonce"] = nonce
    m["time"] = curr_time
    m["expire"] = int(expire)
    m["signature"] = bytes.decode(self.__hmacsha256(roomId, userId, nonce, curr_time, expire))

    raw_sig = json.dumps(m)
    sig_cmpressed = zlib.compress(raw_sig.encode('utf-8'))
    base64_sig = base64_encode_url(sig_cmpressed)
    return base64_sig